canary/webrtc-c/src/media-server-storage/Common.cpp

```diff
@@ -0,0 +1,1864 @@
+        DLOGP("[Signaling connect client] %" PRIu64 " ms", pSampleConfiguration->signalingClientMetrics.signalingClientStats.connectClientTime);
+        if (joinSessionToOfferRecvTime != 0) {
+            DLOGP("[Signaling Join session to offer received] %" PRIu64 " ms", joinSessionToOfferRecvTime);
+        }
+
+        Canary::Cloudwatch::getInstance().monitoring.pushSignalingClientMetrics(&pSampleConfiguration->signalingClientMetrics);
+    }
+
+    if(STATUS_FAILED(peerConnectionGetMetrics(pSampleStreamingSession->pPeerConnection, &pSampleStreamingSession->peerConnectionMetrics))) {
+        DLOGW("Could not get peer connection metrics (0x%08x)", retStatus);
+    } else {
+        Canary::Cloudwatch::getInstance().monitoring.pushPeerConnectionMetrics(&pSampleStreamingSession->peerConnectionMetrics);
+    }
+    if(STATUS_FAILED(iceAgentGetMetrics(pSampleStreamingSession->pPeerConnection, &pSampleStreamingSession->iceMetrics))) {
+        DLOGW("Could not get ICE agent metrics (0x%08x)", retStatus);
+    } else {
+        Canary::Cloudwatch::getInstance().monitoring.pushKvsIceAgentMetrics(&pSampleStreamingSession->iceMetrics);
+    }
+
+CleanUp:
+    return retStatus;
+}
+
+STATUS sendProfilingMetricsTryer(PSampleStreamingSession pSampleStreamingSession)
+{
+    DLOGD("[Canary] sendProfilingMetricsTryer(...) invoked");
+    STATUS retStatus = STATUS_SUCCESS;
+    PSampleConfiguration pSampleConfiguration;
+    BOOL done = FALSE;
+
+    CHK(pSampleStreamingSession != NULL, STATUS_NULL_ARG);
+    pSampleConfiguration = pSampleStreamingSession->pSampleConfiguration;
+    CHK(pSampleConfiguration != NULL, STATUS_NULL_ARG);
+
+    while (!ATOMIC_LOAD_BOOL(&pSampleStreamingSession->terminateFlag) && !ATOMIC_LOAD_BOOL(&pSampleConfiguration->appTerminateFlag)) {
+        retStatus = sendProfilingMetrics(pSampleStreamingSession);
+        if (retStatus == STATUS_WAITING_ON_FIRST_FRAME) {
+            THREAD_SLEEP(HUNDREDS_OF_NANOS_IN_A_MILLISECOND * 100); // to prevent busy waiting
+        } else if (retStatus == STATUS_SUCCESS) {
+            DLOGI("[Canary] First frame sent out, pushed profiling metrics");
+            done = TRUE;
+            break;
+        }
+    }
+
+CleanUp:
+    if(!done) {
+            DLOGE("[Canary] First frame never got sent out...no profiling metrics pushed to cloudwatch..error code: 0x%08x", retStatus);
+        }
+    return retStatus;
+}
+
+STATUS initMetricsTimers(PSampleStreamingSession pSampleStreamingSession)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PSampleConfiguration pSampleConfiguration;
+    
+    CHK(pSampleStreamingSession != NULL, STATUS_NULL_ARG);
+    pSampleConfiguration = pSampleStreamingSession->pSampleConfiguration;
+    CHK(pSampleConfiguration != NULL, STATUS_NULL_ARG);
+
+    pSampleStreamingSession->pushProfilingThread = std::thread(sendProfilingMetricsTryer, pSampleStreamingSession);
+    pSampleStreamingSession->pushProfilingThread.detach();
```

Comment:
By detaching the thread, we don't have to wait on any processes to finish as we would with join(). This thread has a THREAD_SLEEP that `freeSampleStreamingSession()` would at the very minimum have to wait on, and although it is only a 100ms sleep, this would have a significant affect on the `storageDisconnectToFrameSentTime` metric.

Reply:
Why dont we start this thread after the session is set up? The metrics are all post start up right? Is there a reason we need to rush publishing this instead of waiting on P2P set up completion?

Reply:
Ah yes, no need for metric publishing to be initiated here. I moved it to be out of sync with the general flow in the `onConnectionStateChange` callback and switched to using `join`.